nixos/modules/services/cluster/patroni/default.nix

```diff
@@ -0,0 +1,249 @@
+      '';
+    };
+
+    nodeIp = mkOption {
+      type = types.str;
+      example = "192.168.1.1";
+      description = ''
+        IP address of this node.
+      '';
+    };
+
+    otherNodesIps = mkOption {
+      type = types.listOf (types.string);
+      example = [ "192.168.1.2" "192.168.1.3" ];
+      description = ''
+        IP addresses of the other nodes.
+      '';
+    };
+
+    restApiPort = mkOption {
+      type = types.int;
+      default = 8008;
+      description = ''
+        The port on Patroni's REST api listens.
+      '';
+    };
+
+    raft = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        This will configure Patroni to use its own RAFT implementation instead of using a dedicated DCS.
+      '';
+    };
+
+    raftPort = mkOption {
+      type = types.int;
+      default = 5010;
+      description = ''
+        The port on which RAFT listens.
+      '';
+    };
+
+    watchdog = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        This will configure Patroni to use the software watchdog built into the Linux kernel.
+      '';
+    };
+
+    settings = mkOption {
+      type = format.type;
+      default = { };
+      description = ''
+        The primary patroni configuration. See the
+        <link xlink:href="https://patroni.readthedocs.io/en/latest/SETTINGS.html">documentation</link>
+        for possible values.
+        Secrets should be passed in by using the <literal>environment</literal> option.
+      '';
+    };
+
+    environment = mkOption {
+      type = with types; attrsOf (nullOr (oneOf [ str path package ]));
+      default = { };
+      example = {
+        PATRONI_REPLICATION_PASSWORD = "/secret/file";
+        PATRONI_SUPERUSER_PASSWORD = "/secret/file";
+      };
+      description = "Environment variables made available to Patroni, useful for providing secrets from files.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    services.patroni.settings = {
+      scope = cfg.scope;
+      name = cfg.name;
+      namespace = cfg.namespace;
+
+      restapi = {
+        listen = "${cfg.nodeIp}:${toString(cfg.restApiPort)}";
+        connect_address = "${cfg.nodeIp}:${toString(cfg.restApiPort)}";
+      };
+
+      raft = {
+        data_dir = "${cfg.dataDir}/raft";
+        self_addr = "${cfg.nodeIp}:5010";
+        partner_addrs = map (ip: ip + ":5010") cfg.otherNodesIps;
+      };
+
+      postgresql = {
+        listen = "${cfg.nodeIp}:${toString(cfg.postgresqlPort)}";
+        connect_address = "${cfg.nodeIp}:${toString(cfg.postgresqlPort)}";
+        data_dir = cfg.postgresqlDataDir;
+        bin_dir = "${postgresql}/bin";
+        pgpass = "${cfg.dataDir}/pgpass";
+      };
+      watchdog = mkIf cfg.watchdog {
+        mode = "required";
+        device = "/dev/watchdog";
+        safety_margin = 5;
+      };
+    };
+
+    users.users.postgres =
```

Comment:
this ought to be configurable, not hardcoded. reusing the postgres uid also clashes with running a non-HA database on the same host for any reason. `isSystemUser` should be set for service users, shells and homedirs should not be set unless absolutely necessary

Reply:
I basically copied this from the PostgreSQL module.
It would be nice for this to be configurable but at the same time I like the ability to do `sudo -u postgres psql`  (maybe this is why it is done this way in the PostgreSQL module).
Either way I don't mind changing this but I'm not sure what to do exactly. Maybe creating `patroni` as a service user and let people create `postgres` user on their own if they want to?

Reply:
one common solution is to have default settings for a number of service accounts (eg `patroni`, `patroni-postgres`, etc) that are created unless the defaults are changed, and used in the configuration where a user/group is required. for this you probably don't need the fixed uid assignment, so merely adding a few user options to the config and creating them unless changed from the default should be quite good.